File: crates/api/src/dto/catalog/movie.rs
use serde::Serialize;

use domain::catalog::Movie;

use crate::dto::common::{ArtworkDto, ContentRatingDto};

#[derive(Debug, Serialize)]
pub struct MovieResponse {
    pub id: String,
    pub title: String,
    pub year: Option<u16>,
    pub overview: Option<String>,
    pub runtime_minutes: Option<u32>,
    pub content_rating: Option<ContentRatingDto>,
    pub manually_edited: bool,
    pub added_at: String,
    pub updated_at: String,
    pub artwork: ArtworkDto,
}

impl From<Movie> for MovieResponse {
    fn from(m: Movie) -> Self {
        MovieResponse {
            id: m.id.0,
            title: m.title,
            year: m.year,
            overview: m.overview,
            runtime_minutes: m.runtime_minutes,
            content_rating: m.content_rating.map(Into::into),
            manually_edited: m.manually_edited,
            added_at: m.added_at.to_string(),
            updated_at: m.updated_at.to_string(),
            artwork: ArtworkDto::from_refs(m.artwork),
        }
    }
}

File: clients/flutter/lib/model/catalog/movie.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'package:shadowmask/model/common/artwork.dart';
import 'package:shadowmask/model/common/content_rating.dart';

part 'movie.freezed.dart';
part 'movie.g.dart';

@freezed
abstract class Movie with _$Movie {
  const factory Movie({
    required String id,
    required String title,
    int? year,
    String? overview,
    int? runtimeMinutes,
    ContentRating? contentRating,
    @Default(false) bool manuallyEdited,
    String? addedAt,
    String? updatedAt,
    Artwork? artwork,
  }) = _Movie;

  factory Movie.fromJson(Map<String, dynamic> json) => _$MovieFromJson(json);
}
